File: orchestrator/src/k8s.ts
import axios from "axios";
import {
  CoreV1Api,
  KubeConfig,
  type V1Pod,
  type V1Service,
  type V1ContainerPort,
  type V1Probe,
  type V1HTTPGetAction,
} from "@kubernetes/client-node";
import {
  K8S_NAMESPACE,
  BROWSER_IMAGE,
  BROWSER_PORT,
  POD_MEMORY_REQUEST,
  POD_CPU_REQUEST,
  POD_MEMORY_LIMIT,
  POD_CPU_LIMIT,
  IMAGE_PULL_SECRET,
  READINESS_INITIAL_DELAY,
  READINESS_PERIOD_SECONDS,
} from "./config";

/**
 * Kubernetes client (CoreV1Api), initialized from the default environment:
 * - In-cluster when running inside Kubernetes
 * - Local kubeconfig when running out-of-cluster (respects KUBECONFIG, defaults to ~/.kube/config)
 */
const kubeConfig = new KubeConfig();
kubeConfig.loadFromDefault();
export const core: CoreV1Api = kubeConfig.makeApiClient(CoreV1Api);

/**
 * Build the Cluster DNS hostname for a per-session Service.
 */
export function browserServiceHost(sessionId: string): string {
  return `browser-session-${sessionId}.${K8S_NAMESPACE}.svc.cluster.local`;
}

/**
 * Construct the readiness probe for the browser container.
 */
export function makeReadinessProbe(): V1Probe {
  const httpGet: V1HTTPGetAction = {
    path: "/v1/health",
    port: BROWSER_PORT,
    scheme: "HTTP",
  };
  return {
    httpGet,
    initialDelaySeconds: READINESS_INITIAL_DELAY,
    periodSeconds: READINESS_PERIOD_SECONDS,
    timeoutSeconds: 5,
    failureThreshold: 12,
    successThreshold: 1,
  };
}

/**
 * Container ports for the browser container.
 */
export function makeContainerPorts(): V1ContainerPort[] {
  return [
    {
      containerPort: BROWSER_PORT,
      name: "http",
      protocol: "TCP",
    },
  ];
}

/**
 * Deterministic pod name for a session.
 */
export function makeSessionPodName(sessionId: string): string {
  return `browser-session-${sessionId}`;
}

/**
 * Deterministic service name for a session.
 */
export function makeSessionServiceName(sessionId: string): string {
  return `browser-session-${sessionId}`;
}

/**
 * Create a dedicated Pod for a session and return its name.
 */
export async function createSessionPod(sessionId: string): Promise<string> {
  const name = makeSessionPodName(sessionId);

  const pod: V1Pod = {
    apiVersion: "v1",
    kind: "Pod",
    metadata: {
      name,
      namespace: K8S_NAMESPACE,
      labels: {
        app: "browser-session",
        role: "active",
        sessionId,
        podName: name,
      },
    },
    spec: {
      restartPolicy: "Always",
      containers: [
        {
          name: "browser",
          image: BROWSER_IMAGE,
          imagePullPolicy: "IfNotPresent",
          env: [
            { name: "PORT", value: String(BROWSER_PORT) },
            { name: "NODE_ENV", value: "production" },
            { name: "BASE_PATH", value: `/sessions/${sessionId}` },
          ],
          ports: makeContainerPorts(),
          readinessProbe: makeReadinessProbe(),
          resources: {
            requests: { memory: POD_MEMORY_REQUEST, cpu: POD_CPU_REQUEST },
            limits: { memory: POD_MEMORY_LIMIT, cpu: POD_CPU_LIMIT },
          },
        },
      ],
      ...(IMAGE_PULL_SECRET
        ? {
            imagePullSecrets: [{ name: IMAGE_PULL_SECRET }],
          }
        : {}),
    },
  };

  await core.createNamespacedPod({ namespace: K8S_NAMESPACE, body: pod });
  return name;
}

/**
 * Create a ClusterIP Service for a session and return its name.
 * The selector determines which Pod(s) are targeted.
 * Additional annotations can be provided and are merged with steel/* defaults.
 */
export async function createSessionService(
  sessionId: string,
  selector: Record<string, string>,
  annotations?: Record<string, string>,
): Promise<string> {
  const name = makeSessionServiceName(sessionId);
  const mergedAnnotations: Record<string, string> = {
    "steel/sessionId": sessionId,
    ...(selector.podName ? { "steel/targetPodName": selector.podName } : {}),
    ...(annotations || {}),
  };

  const service: V1Service = {
    apiVersion: "v1",
    kind: "Service",
    metadata: {
      name,
      namespace: K8S_NAMESPACE,
      labels: {
        app: "browser-session",
        sessionId,
      },
      annotations: mergedAnnotations,
    },
    spec: {
      type: "ClusterIP",
      selector,
      ports: [
        {
          name: "http",
          port: BROWSER_PORT,
          targetPort: BROWSER_PORT,
          protocol: "TCP",
        },
      ],
    },
  };

  await core.createNamespacedService({
    namespace: K8S_NAMESPACE,
    body: service,
  });
  return name;
}

/**
 * Delete a Service by name, ignoring NotFound errors.
 */
export async function deleteService(name: string): Promise<void> {
  try {
    await core.deleteNamespacedService({ name, namespace: K8S_NAMESPACE });
  } catch (err: any) {
    const code =
      err?.response?.status ??
      err?.statusCode ??
      err?.body?.code ??
      err?.code ??
      undefined;
    if (code === 404) {
      return;
    }
    // Re-throw other errors
    throw err;
  }
}

/**
 * Delete a Pod by name, ignoring NotFound errors.
 */
export async function deletePod(name: string): Promise<void> {
  try {
    await core.deleteNamespacedPod({ name, namespace: K8S_NAMESPACE });
  } catch (err: any) {
    const code =
      err?.response?.status ??
      err?.statusCode ??
      err?.body?.code ??
      err?.code ??
      undefined;
    if (code === 404) {
      return;
    }
    // Re-throw other errors
    throw err;
  }
}

/**
 * Return union of prewarm pods identified by labels and name prefix.
 */
export async function listPrewarmPods(): Promise<V1Pod[]> {
  const res = await core.listNamespacedPod({ namespace: K8S_NAMESPACE });
  const all = res.items || [];

  const unionMap: Record<string, V1Pod> = {};
  for (const p of all) {
    const name = p.metadata?.name ?? "";
    const labels = p.metadata?.labels || {};
    const isLabelMatch =
      labels.app === "browser-session" && labels.role === "prewarm";
    const isPrefixMatch = name.startsWith("browser-prewarm-");
    if (isLabelMatch || isPrefixMatch) {
      unionMap[name] = p;
    }
  }
  return Object.values(unionMap);
}

/**
 * Return union of active session services identified by annotations, labels, and name prefix.
 */
export async function listActiveServices(): Promise<V1Service[]> {
  const res = await core.listNamespacedService({ namespace: K8S_NAMESPACE });
  const all = res.items || [];

  const unionMap: Record<string, V1Service> = {};
  for (const s of all) {
    const name = s.metadata?.name ?? "";
    const labels = s.metadata?.labels || {};
    const annotations = s.metadata?.annotations || {};
    const byAnnotation = Boolean(annotations["steel/sessionId"]);
    const byLabel = labels.app === "browser-session";
    const byPrefix = name.startsWith("browser-session-");

    if (byAnnotation || byLabel || byPrefix) {
      unionMap[name] = s;
    }
  }
  return Object.values(unionMap);
}

/**
 * Check if a Pod is Ready based on its status conditions.
 */
export function isPodReady(pod: V1Pod): boolean {
  const conditions = pod.status?.conditions || [];
  return conditions.some((c) => c.type === "Ready" && c.status === "True");
}

/**
 * Poll the per-session Service readiness endpoint until success or timeout.
 * Targets: http://<service>.<namespace>.svc.cluster.local:<port>/v1/health
 */
export async function waitForServiceReadiness(
  serviceName: string,
  timeoutSeconds: number,
): Promise<void> {
  const fqdn = `${serviceName}.${K8S_NAMESPACE}.svc.cluster.local`;
  const deadline = Date.now() + timeoutSeconds * 1000;
  let lastErr: unknown = null;

  while (Date.now() < deadline) {
    try {
      const url = `http://${fqdn}:${BROWSER_PORT}/v1/health`;
      const res = await axios.get(url, { timeout: 3000 });
      if (res.status >= 200 && res.status < 300) {
        return;
      }
      lastErr = new Error(`Unexpected HTTP ${res.status}`);
    } catch (err) {
      lastErr = err;
    }
    await new Promise((r) => setTimeout(r, 1000));
  }

  throw new Error(
    `Timed out waiting for service ${serviceName} readiness: ${lastErr ? String(lastErr) : "unknown error"}`,
  );
}

File: orchestrator/k8s.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"time"

	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
)

type K8sClient struct {
	clientset    *kubernetes.Clientset
	namespace    string
	browserImage string
	browserPort  int
}

func NewK8sClient(namespace, browserImage string, browserPort int) (*K8sClient, error) {
	config, err := rest.InClusterConfig()
	if err != nil {
		return nil, fmt.Errorf("in-cluster config: %w", err)
	}
	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		return nil, fmt.Errorf("clientset: %w", err)
	}
	return &K8sClient{
		clientset:    clientset,
		namespace:    namespace,
		browserImage: browserImage,
		browserPort:  browserPort,
	}, nil
}

// ─── Session Pods ──────────────────────────────────────────────

// CreateSessionPod creates a dedicated pod for a session.
func (k *K8sClient) CreateSessionPod(ctx context.Context, sessionID string) (string, error) {
	podName := fmt.Sprintf("browser-session-%s", sessionID)
	pod := k.makeBrowserPod(podName, sessionID, "active", fmt.Sprintf("/sessions/%s", sessionID))

	_, err := k.clientset.CoreV1().Pods(k.namespace).Create(ctx, pod, metav1.CreateOptions{})
	if err != nil {
		return "", fmt.Errorf("create pod: %w", err)
	}
	return podName, nil
}

// CreatePrewarmPod creates a warm standby pod with a unique name.
// It returns the pod name.
func (k *K8sClient) CreatePrewarmPod(ctx context.Context) (string, error) {
	podName := fmt.Sprintf("browser-warm-%s", generateShortID())

	pod := k.makeBrowserPod(podName, "", "prewarm", "")
	pod.Labels["prewarm"] = "true"

	_, err := k.clientset.CoreV1().Pods(k.namespace).Create(ctx, pod, metav1.CreateOptions{})
	if err != nil {
		return "", fmt.Errorf("create prewarm pod: %w", err)
	}
	slog.Info("Created prewarm pod", "podName", podName)
	return podName, nil
}

// makeBrowserPod builds a browser pod spec with common defaults.
func (k *K8sClient) makeBrowserPod(name, sessionID, role, basePath string) *corev1.Pod {
	labels := map[string]string{
		"app":  "browser-session",
		"role": role,
	}
	if sessionID != "" {
		labels["sessionId"] = sessionID
	}

	env := []corev1.EnvVar{
		{Name: "PORT", Value: fmt.Sprintf("%d", k.browserPort)},
		{Name: "NODE_ENV", Value: "production"},
		{Name: "CHROME_HEADLESS", Value: "true"},
		{Name: "SKIP_FINGERPRINT_INJECTION", Value: "true"},
	}
	if basePath != "" {
		env = append(env, corev1.EnvVar{Name: "BASE_PATH", Value: basePath})
	}

	return &corev1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      name,
			Namespace: k.namespace,
			Labels:    labels,
		},
		Spec: corev1.PodSpec{
			RestartPolicy: corev1.RestartPolicyAlways,
			Containers: []corev1.Container{
				{
					Name:            "browser",
					Image:           k.browserImage,
					ImagePullPolicy: corev1.PullIfNotPresent,
					Env:             env,
					Ports: []corev1.ContainerPort{
						{ContainerPort: int32(k.browserPort), Name: "http", Protocol: corev1.ProtocolTCP},
					},
					ReadinessProbe: &corev1.Probe{
						ProbeHandler: corev1.ProbeHandler{
							HTTPGet: &corev1.HTTPGetAction{
								Path:   "/v1/health",
								Port:   intstrFromInt(k.browserPort),
								Scheme: corev1.URISchemeHTTP,
							},
						},
						InitialDelaySeconds: 3,
						PeriodSeconds:       5,
						TimeoutSeconds:      5,
						FailureThreshold:    12,
					},
					Resources: corev1.ResourceRequirements{
						Requests: mustParseResources("256Mi", "200m"),
						Limits:   mustParseResources("512Mi", "500m"),
					},
				},
			},
		},
	}
}

// ─── Session Services ──────────────────────────────────────────

func (k *K8sClient) CreateSessionService(ctx context.Context, sessionID string) error {
	svcName := fmt.Sprintf("browser-session-%s", sessionID)

	svc := &corev1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name:      svcName,
			Namespace: k.namespace,
			Labels: map[string]string{
				"app":       "browser-session",
				"sessionId": sessionID,
			},
			Annotations: map[string]string{
				"steel/sessionId": sessionID,
			},
		},
		Spec: corev1.ServiceSpec{
			Type:     corev1.ServiceTypeClusterIP,
			Selector: map[string]string{"sessionId": sessionID},
			Ports: []corev1.ServicePort{
				{
					Name:       "http",
					Port:       int32(k.browserPort),
					TargetPort: intstrFromInt(k.browserPort),
					Protocol:   corev1.ProtocolTCP,
				},
			},
		},
	}

	_, err := k.clientset.CoreV1().Services(k.namespace).Create(ctx, svc, metav1.CreateOptions{})
	if err != nil {
		return fmt.Errorf("create service: %w", err)
	}
	return nil
}

// ─── Deletion ──────────────────────────────────────────────────

func (k *K8sClient) DeleteService(ctx context.Context, name string) {
	_ = k.clientset.CoreV1().Services(k.namespace).Delete(ctx, name, metav1.DeleteOptions{})
}

func (k *K8sClient) DeletePod(ctx context.Context, name string) {
	_ = k.clientset.CoreV1().Pods(k.namespace).Delete(ctx, name, metav1.DeleteOptions{})
}

// ─── Queries ───────────────────────────────────────────────────

type ServiceInfo struct {
	Name        string
	Annotations map[string]string
}

func (k *K8sClient) ListActiveServices(ctx context.Context) ([]ServiceInfo, error) {
	svcs, err := k.clientset.CoreV1().Services(k.namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, err
	}

	var result []ServiceInfo
	for _, svc := range svcs.Items {
		name := svc.Name
		labels := svc.Labels
		annotations := svc.Annotations
		if annotations == nil {
			annotations = map[string]string{}
		}

		isSession := (annotations["steel/sessionId"] != "") ||
			(labels["app"] == "browser-session") ||
			(len(name) > 17 && name[:17] == "browser-session-")

		if isSession && name != "browser-orchestrator" {
			result = append(result, ServiceInfo{
				Name:        name,
				Annotations: annotations,
			})
		}
	}
	return result, nil
}

func (k *K8sClient) GetPodIP(ctx context.Context, podName string) (string, error) {
	pod, err := k.clientset.CoreV1().Pods(k.namespace).Get(ctx, podName, metav1.GetOptions{})
	if err != nil {
		return "", fmt.Errorf("get pod: %w", err)
	}
	if pod.Status.PodIP == "" {
		return "", fmt.Errorf("pod %s has no IP yet", podName)
	}
	return pod.Status.PodIP, nil
}

// ─── Warm Pool Queries ─────────────────────────────────────────

// PrewarmPodInfo describes a warm standby pod.
type PrewarmPodInfo struct {
	Name  string
	Ready bool
	IP    string
}

// ListPrewarmPods returns all prewarm pods that are currently running.
func (k *K8sClient) ListPrewarmPods(ctx context.Context) ([]PrewarmPodInfo, error) {
	pods, err := k.clientset.CoreV1().Pods(k.namespace).List(ctx, metav1.ListOptions{
		LabelSelector: "app=browser-session,role=prewarm",
	})
	if err != nil {
		return nil, err
	}

	var result []PrewarmPodInfo
	for _, pod := range pods.Items {
		if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
			continue
		}
		ready := false
		for _, c := range pod.Status.Conditions {
			if c.Type == corev1.PodReady && c.Status == corev1.ConditionTrue {
				ready = true
			}
		}
		result = append(result, PrewarmPodInfo{
			Name:  pod.Name,
			Ready: ready,
			IP:    pod.Status.PodIP,
		})
	}
	return result, nil
}

// ClaimPrewarmPod takes a ready prewarm pod, relabels it for the given session,
// renames it to the canonical session pod name, and returns the new pod name.
// Since Kubernetes doesn't support renaming pods, we relabel in-place.
func (k *K8sClient) ClaimPrewarmPod(ctx context.Context, sessionID, podName string) error {
	// Relabel the pod: set role=active and add sessionId label
	patchLabels := map[string]interface{}{
		"metadata": map[string]interface{}{
			"labels": map[string]string{
				"role":      "active",
				"sessionId": sessionID,
				"prewarm":   "",
			},
		},
	}
	patchData, _ := json.Marshal(patchLabels)
	_, err := k.clientset.CoreV1().Pods(k.namespace).Patch(
		ctx, podName, "application/strategic-merge-patch+json", patchData, metav1.PatchOptions{},
	)
	if err != nil {
		return fmt.Errorf("patch prewarm pod labels: %w", err)
	}
	slog.Info("Claimed prewarm pod", "podName", podName, "sessionId", sessionID)
	return nil
}

// WaitForReadiness polls the service until it responds healthy or times out.
func (k *K8sClient) WaitForReadiness(ctx context.Context, serviceName string, timeout time.Duration) error {
	fqdn := fmt.Sprintf("%s.%s.svc.cluster.local", serviceName, k.namespace)
	healthURL := fmt.Sprintf("http://%s:%d/v1/health", fqdn, k.browserPort)
	client := &http.Client{Timeout: 3 * time.Second}

	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		resp, err := client.Get(healthURL)
		if err == nil && resp.StatusCode >= 200 && resp.StatusCode < 300 {
			resp.Body.Close()
			return nil
		}
		if resp != nil {
			resp.Body.Close()
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(time.Second):
		}
	}
	return fmt.Errorf("timed out waiting for service %s readiness", serviceName)
}
